--- src/expr/mkexpr.orig	2020-06-19 10:59:27.000000000 -0600
+++ src/expr/mkexpr	2022-10-11 14:28:31.120453409 -0600
@@ -16,6 +16,7 @@
 #
 
 copyright=2010-2014
+shopt -u patsub_replacement
 
 filename=`basename "$1" | sed 's,_template,,'`
 
--- src/expr/mkkind.orig	2020-06-19 10:59:27.000000000 -0600
+++ src/expr/mkkind	2022-10-11 14:34:17.008996126 -0600
@@ -15,6 +15,7 @@
 #
 
 copyright=2010-2014
+shopt -u patsub_replacement
 
 filename=`basename "$1" | sed 's,_template,,'`
 
--- src/expr/mkmetakind.orig	2020-06-19 10:59:27.000000000 -0600
+++ src/expr/mkmetakind	2022-10-11 14:34:32.248020036 -0600
@@ -18,6 +18,7 @@
 #
 
 copyright=2010-2014
+shopt -u patsub_replacement
 
 cat <<EOF
 /*********************                                                        */
