From: Peter Lemenkov <lemenkov@gmail.com>
Date: Fri, 1 Jul 2022 10:14:04 +0200
Subject: [PATCH] Stop checking for jsoncpp version

Signed-off-by: Peter Lemenkov <lemenkov@gmail.com>

diff --git a/libsolutil/JSON.cpp b/libsolutil/JSON.cpp
index ea75ada0f..9de4b16b2 100644
--- a/libsolutil/JSON.cpp
+++ b/libsolutil/JSON.cpp
@@ -30,11 +30,6 @@
 #include <map>
 #include <memory>
 
-static_assert(
-	(JSONCPP_VERSION_MAJOR == 1) && (JSONCPP_VERSION_MINOR == 9) && (JSONCPP_VERSION_PATCH == 3),
-	"Unexpected jsoncpp version: " JSONCPP_VERSION_STRING ". Expecting 1.9.3."
-);
-
 namespace solidity::util
 {
 
