From: Martin Blicha <martin.blicha@gmail.com>
Date: Fri, 2 Jun 2023 13:49:29 +0200
Subject: [PATCH] SMTChecker: Rename CVC interface files


diff --git a/libsmtutil/CMakeLists.txt b/libsmtutil/CMakeLists.txt
index 905c983fa..1363dc4a1 100644
--- a/libsmtutil/CMakeLists.txt
+++ b/libsmtutil/CMakeLists.txt
@@ -19,7 +19,7 @@ else()
 endif()
 
 if (${CVC5_FOUND})
-  set(cvc5_SRCS CVC4Interface.cpp CVC4Interface.h)
+  set(cvc5_SRCS CVC5Interface.cpp CVC5Interface.h)
 else()
   set(cvc5_SRCS)
 endif()
diff --git a/libsmtutil/CVC4Interface.cpp b/libsmtutil/CVC5Interface.cpp
similarity index 99%
rename from libsmtutil/CVC4Interface.cpp
rename to libsmtutil/CVC5Interface.cpp
index 59570fc1d..d55c785cd 100644
--- a/libsmtutil/CVC4Interface.cpp
+++ b/libsmtutil/CVC5Interface.cpp
@@ -16,7 +16,7 @@
 */
 // SPDX-License-Identifier: GPL-3.0
 
-#include <libsmtutil/CVC4Interface.h>
+#include <libsmtutil/CVC5Interface.h>
 
 #include <libsolutil/CommonData.h>
 #include <libsolutil/CommonIO.h>
diff --git a/libsmtutil/CVC4Interface.h b/libsmtutil/CVC5Interface.h
similarity index 100%
rename from libsmtutil/CVC4Interface.h
rename to libsmtutil/CVC5Interface.h
diff --git a/libsmtutil/SMTPortfolio.cpp b/libsmtutil/SMTPortfolio.cpp
index 42aee2b0a..109e1a671 100644
--- a/libsmtutil/SMTPortfolio.cpp
+++ b/libsmtutil/SMTPortfolio.cpp
@@ -22,7 +22,7 @@
 #include <libsmtutil/Z3Interface.h>
 #endif
 #ifdef HAVE_CVC5
-#include <libsmtutil/CVC4Interface.h>
+#include <libsmtutil/CVC5Interface.h>
 #endif
 #include <libsmtutil/SMTLib2Interface.h>
 
